lib/src/git.rs

```diff
@@ -348,6 +349,192 @@ pub fn git_worktree_add(
+    let worktree_data = worktrees_path.join(name);
+    let worktree_data_canon = worktree_data
+        .canonicalize()
+        .map_err(|_| WorktreeRemovalValidationError::NonexistentWorktree(name.to_string()))?;
+
+    let working_directory_state = if base_path.exists() {
+        // If the working directory exists,  we expect .git to exist and be a file
+        //
+        // git fails in this case with
+        //
+        // > fatal: validation failed, cannot remove working tree: '/private/tmp/fourth/.git' does not exist
+        if !dotgit.exists() {
+            return Err(WorktreeRemovalValidationError::MissingDotGit {
+                name: name.to_owned(),
+                dotgit: dotgit.clone(),
+            });
+        }
+        // Now read the .git file and see if it points back to us
+        //
+        // In all these failure modes, git fails with
+        //
+        // > fatal: validation failed, cannot remove working tree: '/private/tmp/fourth/.git' is not a .git file, error code 7
+        let dotgit_content = fs::read_to_string(&dotgit).map_err(|_io_err| {
+            WorktreeRemovalValidationError::BrokenDotGit {
+                name: name.to_owned(),
+                dotgit: dotgit.clone(),
+            }
+        })?;
+
+        let dotgit_gitdir_canon = dotgit_content
+            .strip_prefix("gitdir: ")
+            .map(|rem| rem.trim())
+            .map(Path::new)
+            .and_then(|p| p.canonicalize().ok())
+            .ok_or_else(|| WorktreeRemovalValidationError::BrokenDotGit {
+                name: name.to_owned(),
+                dotgit: dotgit.clone(),
+            })?;
+
+        if worktree_data_canon != dotgit_gitdir_canon {
+            return Err(WorktreeRemovalValidationError::BrokenDotGit {
+                name: name.to_owned(),
+                dotgit: dotgit.clone(),
+            });
+        }
+
+        // Now, we should also delete the .git file. So keep it around for when we do delete this.
+        WorktreeWorkingDirectoryState::Present { dotgit }
+    } else {
+        WorktreeWorkingDirectoryState::Prunable
+    };
+
+    Ok(WorktreeStat {
+        working_directory_state,
+        worktree_data: worktree_data_canon,
+    })
+}
+
+/// `git worktree remove` implementation
+///
+/// Important note when checking functionality against Git itself --
+/// `git worktree remove` is documented as taking a <worktree> parameter (i.e. a
+/// name) but it only works with paths, or maybe the named method only works
+/// when the last path segment of the worktree location matches the worktree
+/// name.
+pub fn git_worktree_remove(stat: WorktreeStat) -> Result<(), WorktreeRemovalValidationError> {
+    let WorktreeStat {
+        working_directory_state,
+        worktree_data,
+    } = stat;
+    fs::remove_dir_all(&worktree_data).map_err(WorktreeRemovalValidationError::RemoveData)?;
+    let worktrees = worktree_data.parent().unwrap();
+    let other_dentries = worktrees
```

Comment:
Nit: Other directories is probably a better name, than dentry.

Reply:
Fixed but no longer part of this batch of changes